### PR TITLE
Use unversioned list options in server

### PR DESCRIPTION
**Pull request by wojtek-t**
Ref #16376

### CODE REVIEW

**Comment by ArtfulCoder** [comment]
@wojtek-t  

rebase needed. do you want to squash the commits ?

**Comment by wojtek-t** [comment]
@ArtfulCoder - this is based on few other PRs - so wil rebase once they are merged.

**Comment by ncdc** [comment]
cc @deads2k @kubernetes/rh-cluster-infra

**Comment by wojtek-t** [comment]
@lavalamp - this is built on top of 3 PRs that you already LGTMed

Can you please take a look at the last commit in this PR?

**Comment by wojtek-t** [comment]
>
Honestly, I think this interface should have a NewOptions() runtime.Object function, and this List function should take an options runtime.Object parameter (which it should expect to have been created by the NewOptions() method).

I think I don't understand this comment. Basically, all REST classes from pkg/registry/*/etcd/etcd.go are implementing those interfaces.
So you would like to extend those interfaces (which means basically extending pkg/registry/generic/etcd/etcd.go) with methods:
NewListOptions()
NewWatchOptions()
both returning *unversioned.ListOptions?

 What is the purpose of that? For me it's overcomplication that doesn't give much any value...

 Also note, that GracefulDeleter is taking explicitly DeleteOptions in its Delete method.

**Comment by wojtek-t** [comment]
@krousey @caesarxuchao - since @lavalamp is on vacation for the whole week, can one of you please take a look on this PR? I know @lavalamp was mostly ok with it, except from the comment from:
https://github.com/kubernetes/kubernetes/pull/17369#issuecomment-158942072

**Comment by caesarxuchao** [comment]
I'll take a look.

**Comment by wojtek-t** [comment]
@caesarxuchao - thanks!

File: pkg/apiserver/resthandler.go [hunk 1]
```diff
@@ -239,31 +239,24 @@ func ListResource(r rest.Lister, rw rest.Watcher, scope RequestScope, forceWatch
...
-		// TODO: Should this be done as part of convertion?
-		fn := func(label, value string) (newLabel, newValue string, err error) {
-			return scope.Convertor.ConvertFieldLabel(scope.APIVersion, scope.Kind, label, value)
-		}
-		if opts.FieldSelector, err = opts.FieldSelector.Transform(fn); err != nil {
-			// TODO: allow bad request to set field causes based on query parameters
-			err = errors.NewBadRequest(err.Error())
-			errorJSON(err, scope.Codec, w)
-			return
+		// TODO: queryToObject should do this.
```
Comment (caesarxuchao) [inline pkg/apiserver/resthandler.go]: queryToObject? Do you mean DecodeParametersInto?

File: pkg/apiserver/apiserver_test.go [hunk 2]
```diff
@@ -931,7 +904,7 @@ func TestList(t *testing.T) {
 			legacy:    true,
 		},
 		{
-			url:       "/" + prefix + "/" + testGroupVersion.Group + "/" + testGroupVersion.Version + "/simple?namespace=other&labels=a%3Db&fields=c%3Dd",
+			url:       "/" + prefix + "/" + testGroupVersion.Group + "/" + testGroupVersion.Version + "/simple?namespace=other&labelSelector=a%3Db&fieldSelector=c%3Dd",
```
Comment (caesarxuchao) [inline pkg/apiserver/apiserver_test.go]: Sorry I'm not familiar with the background of this PR. Is `labels=a%3Db&fields=c%3Dd` a query sent by old clients? If so, we probably should keep the original line to test the backward compatibility.

**Comment by caesarxuchao** [comment]
@wojtek-t could you point me to where lavalamp's comment is about? Perhaps it's a comment on a commit and now it losts its context.

File: pkg/apiserver/apiserver_test.go (hunk 2 above)
Comment (wojtek-t) [inline pkg/apiserver/apiserver_test.go]: No - it's not previous format - it's only the format that is introduced in the test - see ListOptions struct that I'm removing from addTestTypes(). In all production code we have labelSelector & fieldSelector.

**Comment by wojtek-t** [comment]
@caesarxuchao - yes, sorry it was in the commit. See this one for the reference:
https://github.com/wojtek-t/kubernetes/commit/44a8ca5b64599b310cc80c18015214b6caf4e1e2#commitcomment-14535221

File: pkg/apiserver/apiserver_test.go (hunk 2 above)
Comment (caesarxuchao) [inline pkg/apiserver/apiserver_test.go]: Ah, thanks.

**Comment by caesarxuchao** [comment]
>
Honestly, I think this interface should have a NewOptions() runtime.Object function, and this List function should take an options runtime.Object parameter (which it should expect to have been created by the NewOptions() method).

I don't understand the purpose of doing this. I can imagine it will be useful if we are going to use different Options for different resources in the future, is that the case?

**Comment by caesarxuchao** [comment]
Just a nit in the comment, and I don't quite follow lavalamp's comment. Otherwise LGTM.

**Comment by wojtek-t** [comment]
@caesarxuchao - thanks!
I will fix this nit and self-apply lgtm label. I can change that when @lavalamp is back if we decide to do it.

File: pkg/apiserver/resthandler.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/apiserver/resthandler.go]: Done.

**Comment by wojtek-t** [comment]
Ref #15945

**Pull request merged**